sql/core/src/main/scala/org/apache/spark/sql/execution/DriverSortExec.scala

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution
+
+import java.util.concurrent.TimeUnit.NANOSECONDS
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{Attribute, NamedExpression, SortOrder, UnsafeProjection}
+import org.apache.spark.sql.catalyst.expressions.codegen.LazilyGeneratedOrdering
+import org.apache.spark.sql.catalyst.plans.physical.{Partitioning, SinglePartition}
+import org.apache.spark.sql.execution.metric.{SQLMetric, SQLMetrics}
+
+/**
+ * A special operator for global [[SortExec]] if it is the root node with/without [[ProjectExec]].
+ */
+case class DriverSortExec(
```

Comment:
Can we use `TakeOrderedAndProjectExec` to replace it? They seem to be similar. What if we give a large enough limit to `TakeOrderedAndProjectExec`?

Reply:
We have a rule `EliminateLimits` to remove limit if it's large enough, i.e., the actually row count is less than or equal to limit number. So if the `Sort` is the root node that means:
1. the user does not specify the limit
2. the limit has no meaning and has been eliminated

I think if the data size is small enough, `DriverSortExec` is faster than `TakeOrderedAndProjectExec`:
- `DriverSortExec` only sort data once at driver side
- `TakeOrderedAndProjectExec` sort the data at executor side then merge the sorted data at driver side

Reply:
`DriverSortExec` must has a local sort below, isn't it?

Reply:
no, it is the local sort..

Reply:
have you tried to do a local sort first? I think merging sorted data in the driver side should be faster.